src/widget/wstarrating.cpp

```diff
@@ -73,58 +77,60 @@ void WStarRating::mouseMoveEvent(QMouseEvent *event) {
     int star = starAtPosition(event->x());
 #endif
 
-    if (star != m_starRating.starCount() && star != -1) {
-        m_starRating.setStarCount(star);
-        update();
+    if (star != -1) {
+        updateVisualRating(star);
     }
 }
 
 void WStarRating::slotStarsUp(double v) {
-    if (v > 0 && m_currRating < m_starRating.maxStarCount()) {
-        int star = m_currRating + 1;
-        emit ratingChanged(star);
+    if (v > 0 && m_starCount < m_visualStarRating.maxStarCount()) {
+        slotSetRating(m_starCount + 1);
     }
 }
 
 void WStarRating::slotStarsDown(double v) {
-    if (v > 0 && m_currRating > 0) {
-        int star = m_currRating - 1;
-        emit ratingChanged(star);
+    if (v > 0 && m_starCount > StarRating::kMinStarCount) {
+        slotSetRating(m_starCount - 1);
     }
 }
 
 void WStarRating::leaveEvent(QEvent* /*unused*/) {
-    // reset to applied track rating
-    m_starRating.setStarCount(m_currRating);
+    resetVisualRating();
+}
+
+void WStarRating::updateVisualRating(int starCount) {
+    if (starCount == m_visualStarRating.starCount()) {
+        return;
+    }
+    m_visualStarRating.setStarCount(starCount);
     update();
 }
 
 // The method uses basic linear algebra to find out which star is under the cursor.
-int WStarRating::starAtPosition(int x) {
+int WStarRating::starAtPosition(int x) const {
     // If the mouse is very close to the left edge, set 0 stars.
-    if (x < m_starRating.sizeHint().width() * 0.05) {
+    if (x < m_visualStarRating.sizeHint().width() * 0.05) {
         return 0;
     }
-    int star = (x / (m_starRating.sizeHint().width() / m_starRating.maxStarCount())) + 1;
+    int star = (x / (m_visualStarRating.sizeHint().width() / m_visualStarRating.maxStarCount())) + 1;
 
-    if (star <= 0 || star > m_starRating.maxStarCount()) {
+    if (star <= 0 || star > m_visualStarRating.maxStarCount()) {
         return 0;
     }
 
     return star;
 }
 
 void WStarRating::mouseReleaseEvent(QMouseEvent* /*unused*/) {
-    emit ratingChanged(m_starRating.starCount());
+    slotSetRating(m_visualStarRating.starCount());
```

Comment:
This feels wrong: we now store the stars without confirmation, as if it was an update from the owner/mediator. IMO it was correct before: emit the change (don't store it) and wait for the owner to confirm/reject/reset.

Reply:
Result of this implementation is that you can now set stars in the rating widget of an empty deck.